### PR TITLE
Improve TNC&Cobyla

### LINKED ISSUE #1651: Cobyla freezes in 0T1.16rc1
See https://stackoverflow.com/questions/64605271/meta-model-of-the-field-function-openturns-1-16rc1

### PR DESCRIPTION
**Pull request by regislebrun**
Now the algorithms stop as soon as a point with abnormal values is met (inf or NaN). It fixes #1651. 
This PR also improves the use of the verbosity of optimization algorithms in the distribution factories: the user can control it through the verbosity level of the Log class. If Log::INFO is activated the algorithms show their progresses.

### CODE REVIEW

**Review by josephmure — APPROVED** [review]
Great job! Shouldn't this go into the 1.16 branch? If you want to, you can [change the "base" branch](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request) from "master" to "1.16" .

**Comment by regislebrun** [comment]
@josephmure I just followed your advice, and then I see commits coming from @jschueller. I hope I didn't break anything as I already had to restart this work from scratch today...

**Review by sofianehaddad — COMMENTED** [review]
@regislebrun if it is a bug fix, it should be integrated in 1.16. Isn't it?
your branch should start from `1.16rc1` and not `master`

**Comment by josephmure** [comment]
@regislebrun I have saved your branch on my fork: https://github.com/josephmure/openturns/tree/improve_optimization so you don't have to worry about making mistakes. In the worst case scenario, I will be able to submit your PR myself.

Here is what I would do if I were you: first rebase on the last commit before the version switch to 1.17, that is commit 3e96205e3eb3aa3df621e929d29f5b80bae32c04.
```
git rebase -i 3e96205
```
Then I would drop all commits from "Set version number 1.17dev" to "KrigingResult: Drop inputTransformedSample_".
Finally:
```
git push --force
```

**Comment by josephmure** [comment]
If you don't want to do this yourself, I have done it on my branch https://github.com/josephmure/openturns/tree/regis. I have checked that a PR to 1.16 is possible.

**Comment by sofianehaddad** [comment]
@regislebrun you can "pick" my branch

**Comment by josephmure** [comment]
@sofianehaddad your "improve_optimization" branch is identical to my "regis" branch ^^

**Comment by sofianehaddad** [comment]
@josephmure probably

**Comment by regislebrun** [comment]
@sofianehaddad @josephmure done. Thanks for your help! I will try to do the same by myself for PR#1646

File: ChangeLog [hunk 1]
```diff
@@ -99,10 +99,10 @@
  * #1605 (MaximumLikelihoodFactory cannot be used with FittingTest.Kolmogorov)
  * #1624 (The graphs_loglikelihood_contour example has a bug)
  * #1643 (Problem in MaximumDistribution PDF)
+ * #1651 (Cobyla freezes in 0T1.16rc1)
  * #1654 (StationaryFunctionalCovarianceModel::discretize segfaults)
  * #1660 (Cannot extract continuous modes from KLResult when dimension>1)
 
-
```
Comment (jschueller) [inline ChangeLog:105]: you're dropping a line here

**Review by jschueller — APPROVED** [review]
nice, it's best to catch nans at the root, the trapezoidal test is still failing though

**Comment by jschueller** [comment]
a last rebase (to 1.16 branch!) is needed

**Comment by regislebrun** [comment]
@jschueller I changed the reference output of the TrapezoidalFactory test. Now, it fails on my box as expected, but also on circleci linux. It looks like Cobyla generates NaNs only on linux. Do you have any idea on how to fix it?
With the output that makes mingw happy I get on linux:
```
400/809 Test  #909: pyinstallcheck_TrapezoidalFactory_std ......................................***Failed    0.81 sec
WRN - Cobyla went to an abnormal point=[-nan,-nan,-nan,-nan]
--- /home/devel/project/python/test/t_TrapezoidalFactory_std.expout	2020-11-12 12:14:25.315333454 +0000
+++ /home/devel/project/build/python/test/t_TrapezoidalFactory_std.out	2020-11-12 13:07:24.130634274 +0000
@@ -1,9 +1,9 @@
 distribution= class=Trapezoidal name=Trapezoidal dimension=1 a=1 b=2.3 c=4.5 d=5 h=0.322581
-Estimated distribution= class=Trapezoidal name=Trapezoidal dimension=1 a=1.006 b=2.275 c=4.545 d=4.99 h=0.3198
+Estimated distribution= class=Trapezoidal name=Trapezoidal dimension=1 a=1.036 b=2.421 c=4.092 d=5.083 h=0.3498
 Default distribution= Trapezoidal(a = -2, b = -1, c = 1, d = 2)
 Distribution from parameters= Trapezoidal(a = 1, b = 2.3, c = 4.5, d = 5)
 Trapezoidal          = Trapezoidal(a = 1, b = 2.3, c = 4.5, d = 5)
-Estimated trapezoidal= Trapezoidal(a = 1.006, b = 2.275, c = 4.545, d = 4.99)
+Estimated trapezoidal= Trapezoidal(a = 1.036, b = 2.421, c = 4.092, d = 5.083)
 Default trapezoidal= Trapezoidal(a = -2, b = -1, c = 1, d = 2)
 Trapezoidal from parameters= Trapezoidal(a = 1, b = 2.3, c = 4.5, d = 5)
 Estimated distribution= class=Trapezoidal name=Trapezoidal dimension=1 a=-2.22045e-15 b=0 c=0 d=2.22045e-15 h=4.5036e+14
```
and with the output that makes mingw happy I get on mingw:
```
307/491 Test #308: cppcheck_TrapezoidalFactory_std ..................................***Failed    4.57 sec
--- /home/devel/project/lib/test/t_TrapezoidalFactory_std.expout	2020-11-12 12:13:56.147834833 +0000
+++ /home/devel/project/build/lib/test/t_TrapezoidalFactory_std.out	2020-11-12 13:05:26.374682019 +0000
@@ -1,9 +1,9 @@
 Distribution          =class=Trapezoidal name=Trapezoidal dimension=1 a=1 b=2.3 c=4.5 d=5 h=0.322581
-Estimated distribution=class=Trapezoidal name=Trapezoidal dimension=1 a=1.036 b=2.421 c=4.092 d=5.083 h=0.3498
+Estimated distribution=class=Trapezoidal name=Trapezoidal dimension=1 a=1.006 b=2.275 c=4.545 d=4.99 h=0.3198
 Default distribution=class=Trapezoidal name=Trapezoidal dimension=1 a=-2 b=-1 c=1 d=2 h=0.333333
 Distribution from parameters=class=Trapezoidal name=Trapezoidal dimension=1 a=1 b=2.3 c=4.5 d=5 h=0.322581
 Trapezoidal          =class=Trapezoidal name=Trapezoidal dimension=1 a=1 b=2.3 c=4.5 d=5 h=0.322581
-Estimated trapezoidal=class=Trapezoidal name=Trapezoidal dimension=1 a=1.036 b=2.421 c=4.092 d=5.083 h=0.3498
+Estimated trapezoidal=class=Trapezoidal name=Trapezoidal dimension=1 a=1.006 b=2.275 c=4.545 d=4.99 h=0.3198
 Default trapezoidal=class=Trapezoidal name=Trapezoidal dimension=1 a=-2 b=-1 c=1 d=2 h=0.333333
 Trapezoidal from parameters=class=Trapezoidal name=Trapezoidal dimension=1 a=1 b=2.3 c=4.5 d=5 h=0.322581
 Estimated distribution=class=Trapezoidal name=Trapezoidal dimension=1 a=-2.22045e-15 b=0 c=0 d=2.22045e-15 h=4.5036e+14
```
It's not only ci which is circle here...

**Comment by jschueller** [comment]
you can fix it for linux and skip it on windows (add an "if" in python/test/CMakelists for this test)

**Comment by josephmure** [comment]
We really need an `assert_almost_equal` method for distributions... Isn't disabling tests on some platforms a little dangerous?

**Comment by jschueller** [comment]
it'll do

**Pull request merged**